source/slang/slang-lower-to-ir.cpp

```diff
@@ -5387,65 +5406,78 @@ struct DeclLoweringVisitor : DeclVisitor<DeclLoweringVisitor, LoweredValInfo>
         auto subContext = nestedContext.getContext();
 
         // Emit any generics that should wrap the actual type.
-        emitOuterGenerics(subContext, decl, decl);
+        auto outerGeneric = emitOuterGenerics(subContext, decl, decl);
 
-        IRStructType* irStruct = subBuilder->createStructType();
-        addNameHint(context, irStruct, decl);
-        addLinkageDecoration(context, irStruct, decl);
+        IRInst* resultType = nullptr;
+        if (as<AssocTypeDecl>(decl))
+        {
+            resultType = subBuilder->createAssociatedType();
+        }
+        else
+        {
+            resultType = subBuilder->createStructType();
+        }
 
-        subBuilder->setInsertInto(irStruct);
+        addNameHint(context, resultType, decl);
+        addLinkageDecoration(context, resultType, decl);
 
-        // A `struct` that inherits from another `struct` must start
-        // with a member for the direct base type.
-        //
-        for( auto inheritanceDecl : decl->getMembersOfType<InheritanceDecl>() )
+        if (resultType->op == kIROp_StructType)
         {
```

Comment:
The sheer amount of code that is specific to the non-`AssociatedType` case here argues that even *if* we keep going down this road of having an associated type emit as a top-level entity with linkage, we probably want to break it out as a separate `visitAssociatedTypeDecl` method instead of trying to have `visitAggTypeDecl` handle two very different cases.

Reply:
Cleaned up.